File: src/main/java/javasolutions/arraysandhashing/TopKFrequentElements.java
package javasolutions.arraysandhashing;

import com.sun.source.tree.Tree;
import javasolutions.LeetCodeTask;

import java.util.*;

/**
 * Given an integer array nums and an integer k, return the k most frequent elements.
 * You may return the answer in any order.
 *
 * Example 1:
 * Input: nums = [1,1,1,2,2,3], k = 2
 * Output: [1,2]
 *
 * Example 2:
 * Input: nums = [1], k = 1
 * Output: [1]
 */

public class TopKFrequentElements extends LeetCodeTask {

    public TopKFrequentElements() {
        super(347, "Top K Frequent Elements");
    }

    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> calcMap = new HashMap<>();
        List<Integer>[] appearancesBucket = new ArrayList[nums.length + 1];
        int[] result = new int[k];

        for (int n : nums) {
            calcMap.put(n, calcMap.getOrDefault(n, 0) + 1);
        }

        for (int key : calcMap.keySet()) {
            int freq = calcMap.get(key);
            if(appearancesBucket[freq] == null) {
                appearancesBucket[freq] = new ArrayList<>();
            }
            appearancesBucket[freq].add(key);
        }

        for (int i = nums.length, counter = 0; i>=0 && counter < k; i--) {
            if (appearancesBucket[i] != null) {
                for (int val: appearancesBucket[i]) {
                    result[counter++] = val;
                }
            }
        }
        return result;
    }
}

File: src/main/kotlin/kotlinsolutions/arraysandhashing/TopKFrequentElements.kt
package kotlinsolutions.arraysandhashing

import kotlinsolutions.LeetCodeTask
import java.util.*
import kotlin.collections.ArrayDeque
import kotlin.collections.HashMap

/**
 * Given an integer array nums and an integer k, return the k most frequent elements.
 * You may return the answer in any order.
 *
 * Example 1:
 * Input: nums = [1,1,1,2,2,3], k = 2
 * Output: [1,2]
 *
 * Example 2:
 * Input: nums = [1], k = 1
 * Output: [1]
 */

class TopKFrequentElements : LeetCodeTask(347, "Top K Frequent Elements") {

    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val frequencyMap = nums.toList()
            .groupingBy { it }
            .eachCount()
        val buckets = Array<MutableList<Int>>(nums.size + 1) { mutableListOf() }
        frequencyMap.forEach { (k, v) -> buckets[v].add(k) }
        return buckets.flatMap { it }.takeLast(k).toIntArray()
    }

    fun topKFrequentOptimal(nums: IntArray, k: Int): IntArray {
        val frequencyArray: HashMap<Int, Int> = HashMap()
        val bucket = Array<MutableList<Int>>(nums.size + 1) { mutableListOf() }
        val result = mutableListOf<Int>()

        nums.forEach {
            val count = frequencyArray.getOrDefault(it, 0)
            frequencyArray[it] = count + 1
        }
        frequencyArray.forEach { (key, value) ->
            bucket[value].add(key)
        }
        var remainingElements = k
        for (index in bucket.size - 1 downTo 0) {
            if (remainingElements <= 0) {
                break
            }
            if (bucket[index].isNotEmpty()) {
                val elementsToAdd = minOf(remainingElements, bucket[index].size)
                result.addAll(bucket[index].take(elementsToAdd))
                remainingElements -= elementsToAdd
            }
        }
        return result.toIntArray()
    }
}